compiler/pipes/collect-const-vars.cpp

```diff
@@ -10,87 +10,212 @@
+      case op_concat:
+        return Derived::on_concat(v.as<op_concat>());
+      case op_conv_regexp:
+        return Derived::on_conv_regexp(v.as<op_conv_regexp>());
+      case op_func_call:
+        return Derived::on_func_call(v.as<op_func_call>());
+      case op_string_build:
+        return Derived::on_string_build(v.as<op_string_build>());
+      default:
+        return Derived::fallback(v);
+    }
+  }
+
+  static ResultType on_array(VertexAdaptor<op_array> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType on_string(VertexAdaptor<op_string> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType on_concat(VertexAdaptor<op_concat> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType on_conv_regexp(VertexAdaptor<op_conv_regexp> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType on_func_call(VertexAdaptor<op_func_call> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType on_string_build(VertexAdaptor<op_string_build> v) {
+    return Derived::fallback(v);
+  }
+
+  static ResultType fallback(VertexPtr v [[maybe_unused]]) {
+    kphp_assert_msg(false, "Internal error: invalid visitor in CollectConstVars pass!");
+  }
+};
+
+struct IsComposite : public VertexVisitor<IsComposite, bool> {
+  static bool on_array(VertexAdaptor<op_array> v [[maybe_unused]]) {
+    return true;
+  }
+
+  static bool fallback(VertexPtr v [[maybe_unused]]) {
+    return false;
+  }
+};
+
+struct ShouldStoreOnTopDown : public VertexVisitor<ShouldStoreOnTopDown, bool> {
+  static bool on_conv_regexp(VertexAdaptor<op_conv_regexp> v) {
+    const VertexPtr expr = VertexUtil::get_actual_value(v->expr());
+    return vk::any_of_equal(expr->type(), op_string, op_concat, op_string_build);
+  }
+
+  static bool fallback(VertexPtr v) {
+    return vk::any_of_equal(v->type(), op_concat, op_string_build, op_func_call);
+  }
+};
+
+struct ShouldStoreOnBottomUp : public VertexVisitor<ShouldStoreOnBottomUp, bool> {
+  static bool fallback(VertexPtr v) {
+    return vk::any_of_equal(v->type(), op_string, op_array);
+  }
+};
+
+struct NameGenerator : public VertexVisitor<NameGenerator, std::string> {
+  static constexpr auto prefix = "const_var";
+
+  static std::string fallback(VertexPtr v[[maybe_unused]]) {
+    return gen_unique_name(prefix);
+  }
+
+  static std::string on_string(VertexAdaptor<op_string> v) {
+    return gen_const_string_name(v->str_val);
+  }
+
+  static std::string on_array(VertexAdaptor<op_array> v) {
+    if (is_array_suitable_for_hashing(v)) {
+      return gen_const_array_name(v);
+    }
+    return fallback(v);
+  }
+
+  static std::string on_conv_regexp(VertexAdaptor<op_conv_regexp> v) {
+    if (v->expr()->type() == op_string) {
+      return gen_const_regexp_name(v->expr().as<op_string>()->str_val);
+    }
+    return fallback(v);
+  }
+};
+
+struct ProcessBeforeReplace : public VertexVisitor<ProcessBeforeReplace, VertexPtr> {
```

Comment:
why doesn't NameGenerator unwrap op_define_val before generating a name?

Reply:
Because we don't need it. For now, we just unwrap `op_define_val` on bottom-up traversal.